### PR TITLE
Feature: 2.1: Initial impl. of Query Types.

### PR DESCRIPTION
**Pull request by anpete**
Adds Query Types to the type system. Query Types are like Entity Types but do not require a key to be defined and can only be used for querying. Query Types allow for ad-hoc querying (like anonymous types), but are more flexible because they allow mapping configuration to be specified. E.g. ToTable, HasColumnName etc. Many advanced mapping capabilities are supported, too, such as inheritance mapping and navigations (dependent -> principal only).

#3932, #9290

### CODE REVIEW

**Comment by ErikEJ** [comment]
Formerly known a ViewType??

File: src/EFCore/DbView`.cs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    ///         may be evaluated in memory rather than being translated into a database query.
+    ///     </para>
+    ///     <para>
+    ///         <see cref="DbQuery{TQuery}" /> objects are usually obtained from a <see cref="DbQuery{TQuery}" />
+    ///         property on a derived <see cref="DbContext" /> or from the <see cref="DbContext.Set{TQuery}" />
+    ///         method.
+    ///     </para>
+    /// </summary>
+    /// <typeparam name="TQuery"> The type of view being operated on by this view. </typeparam>
+    public abstract class DbQuery<TQuery>
```
Comment (ajcvickers) [inline src/EFCore/DbView`.cs]: >DbQuery [](start = 26, length = 7)

Rename file to match type name

File: src/EFCore/Extensions/EntityTypeExtensions.cs [hunk 2]
```diff
@@ -39,6 +40,18 @@ public static IEnumerable<IEntityType> GetDerivedTypes([NotNull] this IEntityTyp
             }
         }
 
+        /// <summary>
+        ///     Gets all types in the model that derive from a given entity type.
+        /// </summary>
+        /// <param name="entityType"> The base type to find types that derive from. </param>
+        /// <returns> The derived types. </returns>
+        public static bool IsQueryType([NotNull] this IEntityType entityType)
```
Comment (ajcvickers) [inline src/EFCore/Extensions/EntityTypeExtensions.cs]: Copy paste doc error

File: src/EFCore/Internal/InternalDbView.cs [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+using Microsoft.EntityFrameworkCore.Query.Internal;
+using Microsoft.EntityFrameworkCore.Utilities;
+
+namespace Microsoft.EntityFrameworkCore.Internal
+{
+    /// <summary>
+    ///     This API supports the Entity Framework Core infrastructure and is not intended to be used
+    ///     directly from your code. This API may change or be removed in future releases.
+    /// </summary>
+    public class InternalDbQuery<TQuery> :
```
Comment (ajcvickers) [inline src/EFCore/Internal/InternalDbView.cs]: >InternalDbQuery [](start = 17, length = 15)

Rename file

File: src/EFCore/Metadata/Builders/ViewTypeBuilder.cs [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+    /// <summary>
+    ///     <para>
+    ///         Provides a simple API for configuring a query type.
+    ///     </para>
+    ///     <para>
+    ///         Instances of this class are returned from methods when using the <see cref="ModelBuilder" /> API
+    ///         and it is not designed to be directly constructed in your application code.
+    ///     </para>
+    /// </summary>
+    public class QueryTypeBuilder : IInfrastructure<IMutableModel>, IInfrastructure<InternalEntityTypeBuilder>
```
Comment (ajcvickers) [inline src/EFCore/Metadata/Builders/ViewTypeBuilder.cs]: Rename file here and generic

File: src/EFCore/Metadata/Internal/Model.cs [hunk 5]
```diff
@@ -110,6 +110,32 @@ public virtual IEnumerable<EntityType> GetEntityTypes()
...
+        ///     directly from your code. This API may change or be removed in future releases.
+        /// </summary>
+        public virtual EntityType AddQueryType([NotNull] Type type)
+        {
+            Check.NotNull(type, nameof(type));
+
+            var entityType = new EntityType(type, this, ConfigurationSource.Explicit);
+
+            var keyProperty = entityType.AddProperty("__QueryTypeKey__", typeof(int), ConfigurationSource.Convention);
+
```
Comment (ajcvickers) [inline src/EFCore/Metadata/Internal/Model.cs]: What is the reason for adding the fake key here? Does it ever get set to any value, or used in some other way?

File: src/EFCore.Specification.Tests/DatabindingTestBase.cs [hunk 6]
```diff
@@ -450,7 +450,7 @@ public virtual void Adding_entity_to_local_view_that_is_already_in_the_state_man
         [Theory]
         [InlineData(false)]
         [InlineData(true)]
-        public virtual void Adding_entity_to_state_manager_of_different_type_than_local_view_type_has_no_effect_on_local_view(
```
Comment (ajcvickers) [inline src/EFCore.Specification.Tests/DatabindingTestBase.cs]: >Adding_entity_to_state_manager_of_different_type_than_local_view_type_has_no_effect_on_local_view [](start = 28, length = 97)

Inappropriate rename?

**Comment by ajcvickers** [comment]
A few questions:
* Does there need to be more test coverage for different types of queries--joins, etc?
* Can you add a test to WithConstructors... to test creating query types that have parameterized constructors
* Can query types have owned types?

File: src/EFCore/DbView`.cs (hunk 1 above)
Comment (anpete) [reply src/EFCore/DbView`.cs]: The rename was pretty heinous 😄

**Comment by anpete** [comment]
@ErikEJ Yes!

File: src/EFCore/Metadata/Internal/Model.cs (hunk 5 above)
Comment (anpete) [reply src/EFCore/Metadata/Internal/Model.cs]: This may have been an early requirement to work around the validator. Will try and remove.

File: src/EFCore/Metadata/Internal/Model.cs [hunk 7]
```diff
@@ -110,6 +110,32 @@ public virtual IEnumerable<EntityType> GetEntityTypes()
...
             _clrTypeMap[type] = entityType;
+
+            return AddEntityType(entityType);
+        }
+
+        /// <summary>
+        ///     This API supports the Entity Framework Core infrastructure and is not intended to be used
+        ///     directly from your code. This API may change or be removed in future releases.
+        /// </summary>
+        public virtual EntityType AddQueryType([NotNull] Type type)
```
Comment (pmiddleton) [inline src/EFCore/Metadata/Internal/Model.cs]: Can this be added to IMutableModel?  I have been working on table value function support based off your vt-new-* branches and need access to this method.

Comment (anpete) [reply src/EFCore/Metadata/Internal/Model.cs]: @ajcvickers This would be a break.

**Comment by anpete** [comment]
@ajcvickers 

> Does there need to be more test coverage for different types of queries--joins, etc?

Yes. Hoping to get a hand from @maumar on this.

> Can you add a test to WithConstructors... to test creating query types that have parameterized constructors

Done.

> Can query types have owned types?

No. That would imply identity.

File: src/EFCore/Metadata/Internal/Model.cs (hunk 7 above)
Comment (pmiddleton) [reply src/EFCore/Metadata/Internal/Model.cs]: Is the version after 2.1 a breakable change version?  I think some of the other changes for TVF will introduce some breaks as well so it might be better to hold off.  I'm also not sure of the current timeline for 2.1, but it seems to be getting close, and I don't want to force something if I can't make that deadline.

**Comment by maumar** [comment]
I'm planning to test this extensively as soon as I'm done with correlated collections (hopefully one last issue remaining)

**Comment by anpete** [comment]
@maumar Thanks...I think.

**Comment by anpete** [comment]
@ajcvickers Fake key was removed. This made me think about your question on Owned Types. I'm guessing you were getting at nested objects on Query Types? If so, yeah, feels like we should be able to make it work from the query side. cc @AndriySvyryd for insight here 😄

File: src/EFCore/Metadata/Internal/Model.cs (hunk 7 above)
Comment (anpete) [reply src/EFCore/Metadata/Internal/Model.cs]: Not sure at this stage. I would still try to target 2.1 if you think it's feasible. Include the breaks you would like to make and we can deal with them on a case-by-case basis. - We can often put it workarounds in the short term or suggest alternatives.

**Comment by anpete** [comment]
@bricelam Can you take a look at the minor Migrations changes?
@AndriySvyryd Can you take a look at the less minor CF changes?

File: src/EFCore/Infrastructure/Annotatable.cs [hunk 8]
```diff
@@ -57,14 +57,13 @@ public virtual Annotation AddAnnotation(string name, object value)
         /// <returns> The added annotation. </returns>
         protected virtual Annotation AddAnnotation([NotNull] string name, [NotNull] Annotation annotation)
         {
-            var previousLength = _annotations.Value.Count;
-            SetAnnotation(name, annotation);
-
-            if (previousLength == _annotations.Value.Count)
+            if (FindAnnotation(name) != null)
```
Comment (AndriySvyryd) [inline src/EFCore/Infrastructure/Annotatable.cs]: Why would you do a double lookup?

File: src/EFCore/Infrastructure/ModelCustomizer.cs [hunk 9]
```diff
@@ -61,7 +61,14 @@ protected virtual void FindSets([NotNull] ModelBuilder modelBuilder, [NotNull] D
         {
             foreach (var setInfo in Dependencies.SetFinder.FindSets(context))
             {
-                modelBuilder.Entity(setInfo.ClrType);
+                if (!setInfo.IsQueryType)
```
Comment (AndriySvyryd) [inline src/EFCore/Infrastructure/ModelCustomizer.cs]: Invert condition

File: src/EFCore/Infrastructure/ModelValidator.cs [hunk 10]
```diff
@@ -160,7 +160,11 @@ protected virtual void ValidateNonNullPrimaryKeys([NotNull] IModel model)
         {
             Check.NotNull(model, nameof(model));
 
-            var entityTypeWithNullPk = model.GetEntityTypes().FirstOrDefault(et => et.FindPrimaryKey() == null);
+            var entityTypeWithNullPk
+                = model.GetEntityTypes()
+                    .Where(et => !et.IsQueryType())
+                    .FirstOrDefault(et => et.FindPrimaryKey() == null);
```
Comment (AndriySvyryd) [inline src/EFCore/Infrastructure/ModelValidator.cs]: Combine conditions

File: src/EFCore/Infrastructure/Annotatable.cs (hunk 8 above)
Comment (anpete) [reply src/EFCore/Infrastructure/Annotatable.cs]: What happens if conventions change annotations but _annotations.Count is unchanged?

Comment (AndriySvyryd) [reply src/EFCore/Infrastructure/Annotatable.cs]: Conventions use `SetAnnotation` instead of `AddAnnotation`

---
In reply to: [160510838](https://github.com/aspnet/EntityFrameworkCore/pull/10647#discussion_r160510838) [](ancestors = 160510838)

File: src/EFCore/Internal/DbSetInitializer.cs [hunk 11]
```diff
@@ -35,7 +38,11 @@ public virtual void InitializeSets(DbContext context)
         {
             foreach (var setInfo in _setFinder.FindSets(context).Where(p => p.Setter != null))
             {
-                setInfo.Setter.SetClrValue(context, ((IDbSetCache)context).GetOrAddSet(_setSource, setInfo.ClrType));
+                setInfo.Setter.SetClrValue(
+                    context,
+                    !setInfo.IsQueryType
```
Comment (AndriySvyryd) [inline src/EFCore/Internal/DbSetInitializer.cs]: Invert condition

File: src/EFCore/Metadata/Builders/QueryTypeBuilder.cs [hunk 12]
```diff
@@ -0,0 +1,303 @@
...
+            Check.NotEmpty(annotation, nameof(annotation));
+            Check.NotNull(value, nameof(value));
+
+            Builder.HasAnnotation(annotation, value, ConfigurationSource.Explicit);
+
+            return this;
+        }
+
+        /// <summary>
+        ///     Sets the base type of this view in an inheritance hierarchy.
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Builders/QueryTypeBuilder.cs]: Some leftover references to "view" in this file

File: src/EFCore/Metadata/Builders/QueryTypeBuilder`.cs [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+        ///     <paramref name="annotation" /> already exists its value will be updated.
+        /// </summary>
+        /// <param name="annotation"> The key of the annotation to be added or updated. </param>
+        /// <param name="value"> The value to be stored in the annotation. </param>
+        /// <returns> The same typeBuilder instance so that multiple configuration calls can be chained. </returns>
+        public new virtual QueryTypeBuilder<TQuery> HasAnnotation([NotNull] string annotation, [NotNull] object value) =>
+            (QueryTypeBuilder<TQuery>)base.HasAnnotation(annotation, value);
+
+        /// <summary>
+        ///     Sets the base type of this view in an inheritance hierarchy.
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Builders/QueryTypeBuilder`.cs]: And here too

File: src/EFCore/Metadata/Builders/QueryTypeBuilder.cs [hunk 14]
```diff
@@ -0,0 +1,303 @@
...
+        ///     </para>
+        /// </summary>
+        /// <param name="relatedTypeName"> The name of the query type that this relationship targets. </param>
+        /// <param name="navigationName">
+        ///     The name of the reference navigation property on this query type that represents the relationship. If
+        ///     no property is specified, the relationship will be configured without a navigation property on this
+        ///     end.
+        /// </param>
+        /// <returns> An object that can be used to configure the relationship. </returns>
+        public virtual ReferenceNavigationBuilder HasOne(
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Builders/QueryTypeBuilder.cs]: No `HasMany`? yet?

File: src/EFCore/Metadata/Builders/QueryTypeBuilder.cs [hunk 15]
```diff
@@ -0,0 +1,303 @@
...
+            [NotNull] string relatedTypeName,
+            [CanBeNull] string navigationName = null)
+        {
+            Check.NotEmpty(relatedTypeName, nameof(relatedTypeName));
+            Check.NullButNotEmpty(navigationName, nameof(navigationName));
+
+            var relatedEntityType = Builder.Metadata.FindInDefinitionPath(relatedTypeName) ??
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Builders/QueryTypeBuilder.cs]: Will we support weak query types? If not then you can omit the `FindInDefinitionPath` call.

File: src/EFCore/Metadata/Conventions/Internal/ForeignKeyPropertyDiscoveryConvention.cs [hunk 16]
```diff
@@ -55,6 +56,7 @@ public virtual InternalRelationshipBuilder Apply(InternalRelationshipBuilder rel
                 // Try to invert if one to one or can be converted to one to one
                 else if ((foreignKey.IsUnique
                           || foreignKey.PrincipalToDependent == null)
+                         && !foreignKey.DeclaringEntityType.IsQueryType()
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Conventions/Internal/ForeignKeyPropertyDiscoveryConvention.cs]: >foreignKey.DeclaringEntityType.IsQueryType() [](start = 29, length = 44)

Do we support relationships between query types?

File: src/EFCore/Extensions/EntityTypeExtensions.cs [hunk 17]
```diff
@@ -39,6 +40,18 @@ public static IEnumerable<IEntityType> GetDerivedTypes([NotNull] this IEntityTyp
             }
         }
 
+        /// <summary>
+        ///     Determine whether the provided entity type is a query type.
+        /// </summary>
+        /// <param name="entityType"> The entity type. </param>
+        /// <returns> true if the entity type is a query type; otherwise false. </returns>
+        public static bool IsQueryType([NotNull] this IEntityType entityType)
```
Comment (AndriySvyryd) [inline src/EFCore/Extensions/EntityTypeExtensions.cs]: This should be a property on `IEntityType`

File: src/EFCore/Metadata/Internal/InternalEntityTypeBuilder.cs [hunk 18]
```diff
@@ -1420,6 +1420,12 @@ private static IndexBuilderSnapshot DetachIndex(Index indexToDetach)
                 navigationToTarget != null
                 || inverseNavigation != null);
 
+            if (targetEntityTypeBuilder.Metadata.IsQueryType())
+            {
+                throw new InvalidOperationException(
+                    CoreStrings.ErrorNavCannotTargetQueryType(targetEntityTypeBuilder.Metadata.DisplayName()));
+            }
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Internal/InternalEntityTypeBuilder.cs]: Move this to `ForeignKey.Navigation`

File: src/EFCore/Metadata/Internal/InternalModelBuilder.cs [hunk 19]
```diff
@@ -37,19 +37,19 @@ public InternalModelBuilder([NotNull] Model metadata)
...
 
         /// <summary>
         ///     This API supports the Entity Framework Core infrastructure and is not intended to be used
         ///     directly from your code. This API may change or be removed in future releases.
         /// </summary>
         public virtual InternalEntityTypeBuilder Entity(
-            [NotNull] Type type, ConfigurationSource configurationSource)
-            => Entity(new TypeIdentity(type), configurationSource);
+            [NotNull] Type type, ConfigurationSource configurationSource, bool throwOnQuery = false)
+            => Entity(new TypeIdentity(type), configurationSource, throwOnQuery);
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Internal/InternalModelBuilder.cs]: Separate the `throwOnQuery = false` overload as something like `EntityOrQuery`. You can do this as a separate PR if you want to keep the size down.

File: src/EFCore/Infrastructure/Annotatable.cs (hunk 8 above)
Comment (anpete) [reply src/EFCore/Infrastructure/Annotatable.cs]: Discussed offline. Problem happens when a convention removes an annotation during as a result of a different annotation being added.

File: src/EFCore/Metadata/Conventions/Internal/ForeignKeyPropertyDiscoveryConvention.cs (hunk 16 above)
Comment (anpete) [reply src/EFCore/Metadata/Conventions/Internal/ForeignKeyPropertyDiscoveryConvention.cs]: Not currently.

File: src/EFCore/Metadata/Builders/QueryTypeBuilder.cs (hunk 15 above)
Comment (anpete) [reply src/EFCore/Metadata/Builders/QueryTypeBuilder.cs]: Will remove.

File: src/EFCore.Relational/Metadata/Conventions/Internal/DefiningQueryOrExplicitTableConvention.cs [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+            if (entityType.IsQueryType()
+                && annotation?.Value != null)
+            {
+                switch (name)
+                {
+                    case RelationalAnnotationNames.TableName:
+                        entityType.RemoveAnnotation(CoreAnnotationNames.DefiningQuery);
+                        break;
+                    case CoreAnnotationNames.DefiningQuery:
+                        entityType.RemoveAnnotation(RelationalAnnotationNames.TableName);
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Metadata/Conventions/Internal/DefiningQueryOrExplicitTableConvention.cs]: This should only remove annotations added by conventions. Explicit annotations should be removed or trigger an exception in the extension methods themselves.

File: src/EFCore.Relational/Query/ExpressionVisitors/RelationalEntityQueryableExpressionVisitor.cs [hunk 21]
```diff
@@ -252,19 +252,24 @@ var materializer
                                     p,
                                     _querySource),
                             _querySource,
-                            out var typeIndexMap).Compile();
+                            out var typeIndexMap);
+
+                var materializer = materializerExpression.Compile();
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Query/ExpressionVisitors/RelationalEntityQueryableExpressionVisitor.cs]: Why was this separated?

File: test/EFCore.Tests/DbContextTest.cs [hunk 22]
```diff
@@ -621,7 +622,7 @@ public override void DetectChanges(IStateManager stateManager)
             await Assert.ThrowsAsync<ObjectDisposedException>(() => context.FindAsync(typeof(Random), 77));
 
             var methodCount = typeof(DbContext).GetMethods(BindingFlags.Public | BindingFlags.Instance | BindingFlags.DeclaredOnly).Count();
-            var expectedMethodCount = 41;
+            var expectedMethodCount = 42;
```
Comment (AndriySvyryd) [inline test/EFCore.Tests/DbContextTest.cs]: Finally we've reached the perfect API size

File: test/EFCore.Tests/ModelBuilding/ModelBuilderTestBase.cs [hunk 23]
```diff
@@ -134,6 +134,10 @@ public TestModelBuilder HasAnnotation(string annotation, object value)
             public abstract TestModelBuilder Entity<TEntity>(Action<TestEntityTypeBuilder<TEntity>> buildAction)
                 where TEntity : class;
 
+            public virtual QueryTypeBuilder<TQuery> Query<TQuery>()
+                where TQuery : class
+                => ModelBuilder.Query<TQuery>();
```
Comment (AndriySvyryd) [inline test/EFCore.Tests/ModelBuilding/ModelBuilderTestBase.cs]: Don't forget to test the non-generic version

File: src/EFCore.Relational/Query/ExpressionVisitors/RelationalEntityQueryableExpressionVisitor.cs (hunk 21 above)
Comment (anpete) [reply src/EFCore.Relational/Query/ExpressionVisitors/RelationalEntityQueryableExpressionVisitor.cs]: Easier debugging.

File: src/EFCore.Relational/Metadata/Conventions/Internal/DefiningQueryOrExplicitTableConvention.cs (hunk 20 above)
Comment (anpete) [reply src/EFCore.Relational/Metadata/Conventions/Internal/DefiningQueryOrExplicitTableConvention.cs]: Removed convention.

File: src/EFCore/Metadata/Internal/InternalModelBuilder.cs (hunk 19 above)
Comment (smitpatel) [reply src/EFCore/Metadata/Internal/InternalModelBuilder.cs]: @AndriySvyryd - Is this needed? If yes then file an issue.

File: src/EFCore/Query/EntityQueryModelVisitor.cs [hunk 24]
```diff
@@ -579,7 +590,8 @@ var lastTrackingModifier
                     .OfType<TrackingResultOperator>()
                     .LastOrDefault();
 
-            return !(queryModel.GetOutputDataInfo() is StreamedScalarValueInfo)
+            return !_modelExpressionApplyingExpressionVisitor.IsViewTypeQuery
```
Comment (smitpatel) [inline src/EFCore/Query/EntityQueryModelVisitor.cs]: left over of ViewType

File: src/EFCore.Relational/RelationalViewTypeBuilderExtensions.cs [hunk 25]
```diff
@@ -0,0 +1,172 @@
...
+using Microsoft.EntityFrameworkCore.Metadata.Internal;
+using Microsoft.EntityFrameworkCore.Utilities;
+
+// ReSharper disable once CheckNamespace
+namespace Microsoft.EntityFrameworkCore
+{
+    /// <summary>
+    ///     Relational database specific extension methods for <see cref="QueryTypeBuilder" />.
+    /// </summary>
+    public static class RelationalQueryTypeBuilderExtensions
```
Comment (smitpatel) [inline src/EFCore.Relational/RelationalViewTypeBuilderExtensions.cs]: File name mismatch

**Comment by smitpatel** [comment]
> ��Using KoreBuild 2.1.0-preview1-15651

Why this file? :trollface:

---
Refers to: temp.txt:1 in 1325b6e. [](commit_id = 1325b6ec1816c4372b92de16bae4b5e811223a18, deletion_comment = False)

File: src/EFCore/Metadata/Internal/InternalModelBuilder.cs (hunk 19 above)
Comment (AndriySvyryd) [reply src/EFCore/Metadata/Internal/InternalModelBuilder.cs]: https://github.com/aspnet/EntityFrameworkCore/issues/10913
